wandb/sdk/data_types/saved_model.py

```diff
@@ -0,0 +1,329 @@
+        if os.path.isfile(self._path):
+            # If the path is a file, then we can just add it to the artifact,
+            # First checking to see if the artifact already has the file (use the cache)
+            # Else, add it directly, allowing the artifact adder to rename the file deterministically.
+            already_added_path = artifact.get_added_local_path_name(self._path)
+            if already_added_path is not None:
+                json_obj["path"] = already_added_path
+            else:
+                target_path = os.path.join(
+                    ".wb_data", "saved_models", os.path.basename(self._path)
+                )
+                json_obj["path"] = artifact.add_file(self._path, target_path, True).path
+        elif os.path.isdir(self._path):
+            # If the path is a directory, then we need to add all of the files
+            # The directory must be named deterministically based on the contents of the directory,
+            # but the files themselves need to have their name preserved.
+            # FUTURE: Add this functionality to the artifact adder itself
+            file_paths = []
+            for dirpath, _, filenames in os.walk(self._path, topdown=True):
+                for fn in filenames:
+                    file_paths.append(os.path.join(dirpath, fn))
+            dirname = b64_string_to_hex(md5_files_b64(file_paths))[:20]
+            target_path = os.path.join(".wb_data", "saved_models", dirname)
+            artifact.add_dir(self._path, target_path)
+            json_obj["path"] = target_path
+        else:
+            raise ValueError(
+                f"Expected a path to a file or directory, got {self._path}"
+            )
+
+        return json_obj
+
+    def model_obj(self) -> SavedModelObjType:
+        """ Returns the model object.
+        """
+        if self._model_obj is None:
+            assert self._path is not None, "Cannot load model object without path"
+            self._set_obj(self._deserialize(self._path))
+
+        assert self._model_obj is not None, "Model object is None"
+        return self._model_obj
+
+    # Methods to be implemented by subclasses
+    @staticmethod
+    def _deserialize(path: str) -> SavedModelObjType:
+        """Returns the model object from a path. Allowed to throw errors
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _validate_obj(obj: Any) -> bool:
+        """Validates the model object. Allowed to throw errors
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _serialize(obj: SavedModelObjType, dir_or_file_path: str) -> None:
+        """Save the model to disk. The method will receive a directory path which all
+        files needed for deserialization should be saved. A directory will always be passed if
+        _path_extension is an empty string, else a single file will be passed. Allowed to throw errors
+        """
```

Comment:
Nit: I'm not sure we gain something by making the distinction between a path being a directory or a single file? We can simplify by just always making it a directory.

Reply:
I tried to think of a way around this, but couldn't see a clean way of doing so. The main problem is that if the user provides a path to a file, let's say `my_model.pt`, we would need to copy it to a temp directory before passing to the loader. At this point the loader would need to scan the directory and take action on the single member of the directory. This felt awkward to charge the subclass implementer to do this file-level manipulation.